xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-templates/src/main/resources/templates/register_macros.vm

```diff
@@ -158,12 +162,67 @@ $xwiki.get('ssfx').use('uicomponents/widgets/validation/livevalidation.css', tru
+        ## support icons.
+        #define ($validationContainer)
+          ## Add the proper classes to the validation string if it's a failure on initialisation
+          #if ($field.error &&
+            $field.error == $!validation.failureMessage &&
+            $!failureMessageParams.get("class")!='')
+            #set($extraValidationClass = $extraValidationClass + ' ' + $!failureMessageParams.get("class"))
+          #elseif ($field.params.value)
+            #set($extraValidationClass = $extraValidationClass + ' LV_valid')##
           #end
-          >$field.error</span>
+          <span class="LV_validation_message $extraValidationClass" aria-live="polite" ##
+            ## Add the proper attributes to the validation string if it's a failure on initialisation
+            #if ($field.error &&
+              $field.error == $!validation.failureMessage)
+              #foreach ($entry in $failureMessageParams.entrySet())
+                #if ($entry.key!='class')
+                  $entry.key="$entry.value" ##
+                #end
+              #end
+            #end>
+            <span class="LV_validation_message_valid_icon">$services.icon.renderHTML('accept')</span> ##
+            <span class="LV_validation_message_invalid_icon">$services.icon.renderHTML('cross')</span> ##
+            ## Add the proper content to the validation string if it's a failure on initialisation
+            ## If it's a regex validation, we also want to make sure to keep it shown at all times, we still put the
+            ## message. Note that the class displaying the state of this validation will change though.
+            #if($field.error && $field.error == $!validation.failureMessage ||
+              $!extraValidationClass.toString().contains('regex'))
+              $!validation.failureMessage
+            #end
+          </span>
+        #end
+        #set($extraValidationClass = '')
+        #if ($field.validate.mandatory)
+          #set($extraValidationClass = 'mandatory')
+          #set($validation = $field.validate.mandatory)
+          $validationContainer
+        #end
+        #if ($field.validate.mustMatch)
+          #set($extraValidationClass = 'must-match')
+          #set($validation = $field.validate.mustMatch)
+          $validationContainer
+        #end
+        #if ($field.validate.programmaticValidation)
+          #set($extraValidationClass = 'programmatic-validation')
+          #set($validation = $field.validate.programmaticValidation)
+          $validationContainer
+        #end
+        #if ($field.validate.regex)
+          #set($extraValidationClass = 'regex')
+          #set($validation = $field.validate.regex)
+          $validationContainer
+        #end
+        #foreach ($regex in $field.validate.regexes)
+          #set($extraValidationClass = 'regex-'+ $foreach.count)
+          #set($validation = $regex)
+          $validationContainer
```

Comment:
Honestly it's not easy to review all this, I really think at a point you should get rid of all that and introduce a proper java API for this.

Reply:
Yeah, it feels a lot like trying to make sure the patchwork does not have any holes left...
I tried to cover a lot of validation use cases in the demo video as a pseudo proof that this system works.